File: withdraw/withdraw.js
// Copyright (c) 2024, ARD and contributors
// For license information, please see license.txt

frappe.ui.form.on('Withdraw', {
	setup:function(frm){
		// numpad.init();
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="quarter"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="half"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="one"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="five"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="ten"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="twenty"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="fifty"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="hundred"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="thouthnd"]')[0] });
		// numpad.attach({ target: document.querySelectorAll('[data-fieldname="fivehundred"]')[0] });
	},
	refresh: function(frm) {
		frm.set_query('debit', function(doc) {
			return {
				query: "carat.carat.doctype.withdraw.withdraw.query_fitler_for_account",
				filters: {
					currency: frm.doc.currency,
					customer_group: frm.doc.customer_group,
					
				}
			};
		});
	}
	,customer: function(frm){
		frm.save()
	},
	calc_money:function(frm){
		let quarter = frm.doc.quarter *.25
		let half = frm.doc.half *.5
		let one = frm.doc.one * 1;
		let five = frm.doc.five * 5;
		let ten = frm.doc.ten * 10;
		let twenty = frm.doc.twenty * 20;
		let fifty = frm.doc.fifty * 50;
		let hundred = frm.doc.hundred * 100;
		let thouthnd = frm.doc.custom_1000 * 1000;
		let fivehundred = frm.doc.custom_500 * 500;  
		let tht = frm.doc.custom_2000 * 2000; 
		let tht5 = frm.doc.custom_5000 * 5000;
		
		let total = quarter + half + one + five + ten + twenty + fifty + hundred + thouthnd + fivehundred +tht +tht5
		frm.set_value("amount",total)
	//	frm.refresh(total)
	//},
	//    count_money:function(frm){
	// 	 let quarter = frm.doc.v_quarter *.25
	// 	 let half = frm.doc.v_half *.5
	// 	 let one = frm.doc.v_one * 1;
	// 	 let five = frm.doc.v_five * 5;
	// 	 let ten = frm.doc.v_ten * 10;
	// 	 let twenty = frm.doc.v_twenty * 20;
	// 	 let fifty = frm.doc.v_fifty * 50;
	// 	 let hundred = frm.doc.v_hundred * 100;
	// 	 let thouthnd = frm.doc.v_thouthnd * 1000;
	// 	 let fivehundred = frm.doc.v_fivehundred * 500;
	// 	 let total_approved_amount = quarter + half + one + five + ten + twenty + fifty + hundred + thouthnd + fivehundred 
	// 	 frm.set_value("amount", total_approved_amount)
		 // let total_damaged = frm.doc.amount - total_approved_amount
		 // frm.set_value("total_damaged", total_damaged)

	},quarter: function(frm){
		frm.trigger("calc_money")
	},half: function(frm){
		frm.trigger("calc_money")
	},one: function(frm){
		frm.trigger("calc_money")
	},five: function(frm){
		frm.trigger("calc_money")
	},ten: function(frm){
		frm.trigger("calc_money")
	},twenty: function(frm){
		frm.trigger("calc_money")
	},fifty: function(frm){
		frm.trigger("calc_money")
	},hundred: function(frm){
		frm.trigger("calc_money")
	},custom_500: function(frm){
		frm.trigger("calc_money")
	},custom_1000: function(frm){
		frm.trigger("calc_money")
	},custom_2000: function(frm){
		frm.trigger("calc_money")
	},custom_5000: function(frm){
		frm.trigger("calc_money")







	},
	// v_quarter: function(frm){
	// 	frm.trigger("count_money")
	// },v_half: function(frm){
	// 	frm.trigger("count_money")
	// },v_one: function(frm){
	// 	frm.trigger("count_money")
	// },v_five: function(frm){
	// 	frm.trigger("count_money")
	// },v_ten: function(frm){
	// 	frm.trigger("count_money")
	// },v_twenty: function(frm){
	// 	frm.trigger("count_money")
	// },v_fifty: function(frm){
	// 	frm.trigger("count_money")
	// },v_hundred: function(frm){
	// 	frm.trigger("count_money")
	// },v_thouthnd: function(frm){
	// 	frm.trigger("count_money")
	// },v_fivehundred: function(frm){
	// 	frm.trigger("count_money")
	// }
});

File: withdraw/withdraw.py
# Copyright (c) 2024, ARD and contributors
# For license information, please see license.txt

import frappe
import json
from frappe.model.document import Document
from datetime import date
from frappe import _, msgprint
from frappe.utils import cint, today, flt
from carat.api import make_journal_entry, make_journal_entry_closing


class Withdraw(Document):
    def validate(self):
        # self.v_quarter = flt(self.quarter)
        # self.v_half = flt(self.half)
        # self.v_one = self.one
        # self.v_five = self.five
        # self.v_ten = self.ten
        # self.v_twenty = self.twenty
        # self.v_fifty = self.fifty
        # self.v_hundred = self.hundred
        # self.v_twohundred = self.twohundred
        # self.v_fivehundred = self.fivehundred

        quarter = flt(self.quarter) * 0.25
        half = flt(self.half) * 0.5
        one = self.one * 1
        five = self.five * 5
        ten = self.ten * 10
        twenty = self.twenty * 20
        fifty = self.fifty * 50
        hundred = self.hundred * 100
        twohundred = self.twohundred * 200
        fivehundred = self.custom_500 * 500
        custom_1000 = self.custom_1000 * 1000
        total = quarter + half + one + five + ten + twenty + fifty + hundred + twohundred + fivehundred + custom_1000
        self.amount = total
        # self.total_damaged = self.amount - self.total_approved_amount
        result = []
        settings_doc = frappe.get_single("Carat Settings")
        deposit_table = settings_doc.deposit
        for row in deposit_table:
            if row.currency == self.currency and row.customer_group == self.customer_group:
                result.append(str(row.debit))
        if len(result) == 1:
            self.debit = result[0]
            self.multi = 0
        elif len(result) > 1:
            self.multi = 1
            if not self.debit:
                self.debit = result[0]
        elif len(result) < 1:
            frappe.throw("This transaction is not supported No info in Carat Settings about it, checkout with administrator")
        return result

    def on_submit(self):
        if self.amount == 0:
            frappe.throw("Amount Should Not Equal Zero")
        amount = 0
        rate = 0
        commission_amount = 0
        settings_doc = frappe.get_single("Carat Settings")
        deposit_table = settings_doc.withdraw
        for row in deposit_table:
            if row.currency == self.currency and row.customer_group == self.customer_group:
                debit = row.debit
                credit = row.credit
                commission_account = row.commission_account
                commission_type = row.commission_type
                amount = row.amount
                rate = row.rate
                if commission_type == "Rate":
                    commission_amount = flt(self.amount) * flt(rate)
                if commission_type == "Amount":
                    commission_amount = amount
                make_journal_entry(self, debit, credit, commission_account, commission_amount,transaction_type="withdraw")

    # def before_update_after_submit(self):
    #     if self.workflow_state == 'Counted and Approved':
    #         settings_doc = frappe.get_single("Carat Settings")
    #         deposit_table = settings_doc.deposit
    #         for row in deposit_table:
    #             if row.currency == self.currency and row.customer_group == self.customer_group and row.debit == self.debit:
    #                 debit = row.debit
    #                 credit = row.credit
    #                 main_treasury = row.main_treasury
    #                 damaged_cash_treasury = row.damaged_cash_treasury
    #                 holding_treasury = row.holding_treasury
    #                 make_journal_entry_closing(self, debit, credit, main_treasury,
    #                                             damaged_cash_treasury, holding_treasury,
    #                                               transaction_type="withdraw")
@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def query_fitler_for_account(doctype, txt, searchfield, start, page_len, filters):
    return frappe.db.sql(
        """select debit from `tabWithdraw Settings table`
        where customer_group = %s and currency = %s
        and %s like %s order by name limit %s offset %s"""
        % ("%s","%s", searchfield, "%s", "%s", "%s"),
        (filters["customer_group"], filters["currency"], "%%%s%%" % txt, page_len, start),
        as_list=1,
    )

